### PR TITLE
cosmic-comp: optional XWayland and Systemd

### PR DESCRIPTION
**Pull request by Fl1tzi**
## Description of changes

Two new optionals:

- useXWayland
- useSystemd

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/co/cosmic-comp/package.nix [hunk 1]
```diff
@@ -55,7 +71,7 @@ rustPlatform.buildRustPackage {
 
   # These libraries are only used by the X11 backend, which will not
   # be the common case, so just make them available, don't link them.
-  postInstall = ''
+  postInstall = lib.optionalString useXWayland ''
     wrapProgram $out/bin/cosmic-comp \
       --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [
         xorg.libX11 xorg.libXcursor xorg.libXi xorg.libXrandr
```
Comment (alyssais) [inline pkgs/by-name/co/cosmic-comp/package.nix:77]: These are unrelated to Xwayland — they're for running the compositor itself as an X11 window.  So they shouldn't be behind the same option.

Comment (Fl1tzi) [reply pkgs/by-name/co/cosmic-comp/package.nix]: fixed

File: pkgs/by-name/co/cosmic-comp/package.nix [hunk 2]
```diff
@@ -41,7 +45,19 @@ rustPlatform.buildRustPackage {
   separateDebugInfo = true;
 
   nativeBuildInputs = [ makeBinaryWrapper pkg-config ];
-  buildInputs = [ libglvnd libinput libxkbcommon mesa seatd udev wayland ];
+  buildInputs =
+    [
```
Comment (alyssais) [inline pkgs/by-name/co/cosmic-comp/package.nix:49]: This is very unusual style for Nixpkgs.  Usually it'd be `buildInputs = [`.

File: pkgs/by-name/co/cosmic-comp/package.nix [hunk 3]
```diff
@@ -57,10 +73,16 @@ rustPlatform.buildRustPackage {
...
-      ]} \
-      --prefix PATH : ${lib.makeBinPath [ xwayland ]}
+      --prefix LD_LIBRARY_PATH : ${
+        lib.makeLibraryPath [
+          xorg.libX11
+          xorg.libXcursor
+          xorg.libXi
+          xorg.libXrandr
+        ]
+      } \
```
Comment (alyssais) [inline pkgs/by-name/co/cosmic-comp/package.nix:83]: We don't know what line comes after this if we don't include the useXwayland stuff, so some unexpected line might end up getting added on to this one.

We could unconditionally insert a blank line after (or just in the non-Xwayland case), but it might be better to do what lots of other packages do and build up a shell array for wrapProgram args.

File: pkgs/by-name/co/cosmic-comp/package.nix [hunk 4]
```diff
@@ -41,7 +45,19 @@ rustPlatform.buildRustPackage {
...
+  buildInputs =
+    [
+      libglvnd
+      libinput
+      libxkbcommon
+      mesa
+      seatd
+      udev
+      wayland
+    ] ++ lib.optional useSystemd systemd;
```
Comment (alyssais) [inline pkgs/by-name/co/cosmic-comp/package.nix:57]: What does this do?  We didn't have an explicit systemd dependency before.

**Review by alyssais — CHANGES_REQUESTED** [review]

File: pkgs/by-name/co/cosmic-comp/package.nix (hunk 4 above)
Comment (alyssais) [reply pkgs/by-name/co/cosmic-comp/package.nix:57]: Ah, I see, it's used for readiness notification and some other things.  Makes sense.

File: pkgs/by-name/co/cosmic-comp/package.nix (hunk 3 above)
Comment (Fl1tzi) [reply pkgs/by-name/co/cosmic-comp/package.nix:83]: Do you have a reference implementation of this? I'm not yet aware of this method.

Comment (alyssais) [reply pkgs/by-name/co/cosmic-comp/package.nix:83]: It'd look something like this:

```nix
''
wrapProgramArgs=(--prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [
  xorg.libX11 xorg.libXcursor xorg.libXi xorg.libXrandr
]})
'' + lib.optionalString useXWayland ''
wrapProgramArgs+=(--prefix PATH : ${lib.makeBinPath [ xwayland ]})
'' + ''
wrapProgram $out/bin/cosmic-comp "${wrapProgramArgs[@]}"
''
```

File: pkgs/by-name/co/cosmic-comp/package.nix [hunk 5]
```diff
@@ -56,11 +71,13 @@ rustPlatform.buildRustPackage {
...
-      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [
+    wrapProgramArgs=(--prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [
         xorg.libX11 xorg.libXcursor xorg.libXi xorg.libXrandr
-      ]} \
-      --prefix PATH : ${lib.makeBinPath [ xwayland ]}
+    ]})
+  '' + lib.optionalString useXWayland ''
+    wrapProgramArgs+=(--prefix PATH : ${lib.makeBinPath [ xwayland ]})
+  '' + ''
+    wrapProgram $out/bin/cosmic-comp ''${wrapProgramArgs[@]}
```
Comment (alyssais) [inline pkgs/by-name/co/cosmic-comp/package.nix:80]: I think this needs to be double quoted [to avoid having the shell reinterpret the arguments](https://www.shellcheck.net/wiki/SC2068)

```suggestion
    wrapProgram $out/bin/cosmic-comp "''${wrapProgramArgs[@]}"
```

**Review by alyssais — APPROVED** [review]

**Comment by alyssais** [comment]
Thank you!
